File: extensions/libdepwin/nappgui/src/inet/osx/oshttpreq.m
/*
 * NAppGUI Cross-platform C SDK
 * 2015-2024 Francisco Garcia Collado
 * MIT Licence
 * https://nappgui.com/en/legal/license.html
 *
 * File: oshttpreq.m
 *
 */

/* HTTP request (Cocoa-based implementation) */

#include "oshttpreq.inl"
#include <sewer/nowarn.hxx>
#include <Cocoa/Cocoa.h>
#include <sewer/warn.hxx>
#include <core/arrst.h>
#include <core/heap.h>
#include <core/stream.h>
#include <core/strings.h>
#include <osbs/bmutex.h>
#include <sewer/cassert.h>
#include <sewer/ptr.h>

#if !defined(__MACOS__)
#error This file is only for OSX
#endif

/*---------------------------------------------------------------------------*/

#if defined(MAC_OS_X_VERSION_10_9) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_9
@interface OSXURLDelegate : NSObject <NSURLSessionDelegate, NSURLSessionDataDelegate, NSURLSessionTaskDelegate>
{
  @public
    OSHttp *http;
    bool_t auto_redirect;
}

@end

#endif

/*---------------------------------------------------------------------------*/

struct _oshttp_t
{
    String *host_url;
    uint16_t port;
    bool_t secure;
    bool_t response;

#if defined(MAC_OS_X_VERSION_10_9) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_9
    Mutex *mutex;
    NSURLSession *session;
    OSXURLDelegate *delegate;
#endif

    NSMutableURLRequest *request;
    ierror_t error;
    String *protocol;
    Stream *headers;
    Stream *body;
};

/*---------------------------------------------------------------------------*/

#if defined(MAC_OS_X_VERSION_10_9) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_9

@implementation OSXURLDelegate

#if defined(MAC_OS_X_VERSION_10_12) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_12

- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didFinishCollectingMetrics:(NSURLSessionTaskMetrics *)metrics
{
    NSArray *ar = [metrics transactionMetrics];

    for (NSURLSessionTaskTransactionMetrics *tr in ar)
    {
        NSString *str = [tr networkProtocolName];
        if (str != nil)
        {
            bmutex_lock(self->http->mutex);
            if (self->http->protocol != NULL)
                self->http->protocol = str_c((const char_t *)[str UTF8String]);
            bmutex_unlock(self->http->mutex);
        }
    }

    cassert(metrics != nil);
}

/*---------------------------------------------------------------------------*/

- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task
    willPerformHTTPRedirection:(NSHTTPURLResponse *)response
                    newRequest:(NSURLRequest *)request
             completionHandler:(void (^)(NSURLRequest *_Nullable))completionHandler
{
    unref(session);
    unref(task);
    unref(response);
    if (auto_redirect == TRUE)
        completionHandler(request);
    else
        completionHandler(nil);
}

#endif

@end

#endif

/*---------------------------------------------------------------------------*/

void oshttp_init(void)
{
}

/*---------------------------------------------------------------------------*/

void oshttp_finish(void)
{
}

/*---------------------------------------------------------------------------*/

OSHttp *oshttp_create(const char_t *host, const uint16_t port, const bool_t secure)
{
    OSHttp *http = heap_new(OSHttp);

    if (secure == TRUE)
        http->host_url = str_printf("https://%s:%d", host, port);
    else
        http->host_url = str_printf("http://%s:%d", host, port);

    http->port = port;
    http->secure = secure;

#if defined(MAC_OS_X_VERSION_10_9) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_9
    http->delegate = [[OSXURLDelegate alloc] init];
    http->delegate->http = http;
    http->delegate->auto_redirect = TRUE;
    http->session = [[NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration] delegate:http->delegate delegateQueue:nil] retain];
    http->mutex = bmutex_create();
#endif

    http->request = [[NSMutableURLRequest alloc] init];
    http->response = FALSE;
    http->protocol = NULL;
    http->headers = NULL;
    http->body = NULL;
    return http;
}

/*---------------------------------------------------------------------------*/

void oshttp_destroy(OSHttp **http)
{
    cassert_no_null(http);
    cassert_no_null(*http);
    str_destroy(&(*http)->host_url);

#if defined(MAC_OS_X_VERSION_10_9) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_9
    [(*http)->session release];
    [(*http)->delegate release];
    bmutex_close(&(*http)->mutex);
#endif

    [(*http)->request release];
    str_destopt(&(*http)->protocol);

    if ((*http)->headers != NULL)
        stm_close(&(*http)->headers);

    if ((*http)->body != NULL)
        stm_close(&(*http)->body);

    heap_delete(http, OSHttp);
}

/*---------------------------------------------------------------------------*/

static __INLINE bool_t i_reserved_header(const char_t *header)
{
    if (str_equ_nocase(header, "content-length") == TRUE)
        return TRUE;

    if (str_equ_nocase(header, "authorization") == TRUE)
        return TRUE;

    if (str_equ_nocase(header, "connection") == TRUE)
        return TRUE;

    if (str_equ_nocase(header, "host") == TRUE)
        return TRUE;

    if (str_equ_nocase(header, "proxy-authenticate") == TRUE)
        return TRUE;

    if (str_equ_nocase(header, "proxy-authorization") == TRUE)
        return TRUE;

    if (str_equ_nocase(header, "www-authenticate") == TRUE)
        return TRUE;

    return FALSE;
}

/*---------------------------------------------------------------------------*/

void oshttp_clear_headers(OSHttp *http)
{
    cassert_no_null(http);
    unref(http);
}

/*---------------------------------------------------------------------------*/

void oshttp_add_header(OSHttp *http, const char_t *name, const char_t *value)
{
    if (http->request != nil)
    {
        if (i_reserved_header(name) == FALSE)
        {
            NSString *lname = [NSString stringWithUTF8String:name];
            NSString *lvalue = [NSString stringWithUTF8String:value];
            [http->request setValue:lvalue forHTTPHeaderField:lname];
        }
    }
}

/*---------------------------------------------------------------------------*/

static ierror_t i_http_error(const NSInteger code)
{
    switch (code)
    {
    case NSURLErrorUnknown:
        return ekIUNDEF;
    case NSURLErrorCancelled:
        return ekIUNDEF;
    case NSURLErrorBadURL:
        return ekINOHOST;
    case NSURLErrorTimedOut:
        return ekITIMEOUT;
    case NSURLErrorUnsupportedURL:
        return ekINOHOST;
    case NSURLErrorCannotFindHost:
        return ekINOHOST;
    case NSURLErrorCannotConnectToHost:
        return ekINOHOST;
    case NSURLErrorNetworkConnectionLost:
        return ekINONET;
    case NSURLErrorDNSLookupFailed:
        return ekINOHOST;
    case NSURLErrorHTTPTooManyRedirects:
        return ekINOHOST;
    case NSURLErrorResourceUnavailable:
        return ekINOHOST;
    case NSURLErrorNotConnectedToInternet:
        return ekINONET;
    case NSURLErrorRedirectToNonExistentLocation:
        return ekINOHOST;
    case NSURLErrorBadServerResponse:
        return ekISERVER;
    case NSURLErrorUserCancelledAuthentication:
        return ekISERVER;
    case NSURLErrorUserAuthenticationRequired:
        return ekISERVER;
    case NSURLErrorZeroByteResource:
        return ekISERVER;
    case NSURLErrorCannotDecodeRawData:
        return ekISERVER;
    case NSURLErrorCannotDecodeContentData:
        return ekISERVER;
    case NSURLErrorCannotParseResponse:
        return ekISERVER;
    case NSURLErrorFileDoesNotExist:
        return ekINOHOST;
    case NSURLErrorFileIsDirectory:
        return ekISERVER;
    case NSURLErrorNoPermissionsToReadFile:
        return ekISERVER;
    default:
        return ekIUNDEF;
    }
}

/*---------------------------------------------------------------------------*/

static void i_response(NSData *data, NSURLResponse *response, NSError *error, OSHttp *http)
{
#if defined(MAC_OS_X_VERSION_10_9) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_9
    bmutex_lock(http->mutex);
#endif

    cassert(http->response == FALSE);
    cassert(http->headers == NULL);
    cassert(http->body == NULL);

    if (error != nil)
    {
        http->response = TRUE;
        http->error = i_http_error([error code]);
    }
    else if (data != nil)
    {
        NSHTTPURLResponse *http_response = (NSHTTPURLResponse *)response;
        NSInteger code = [http_response statusCode];
        NSDictionary *rheaders = [http_response allHeaderFields];
        NSString *rmessage = [NSHTTPURLResponse localizedStringForStatusCode:code];
        uint32_t dsize = (uint32_t)[data length];
        const byte_t *dbytes = [data bytes];

        http->headers = stm_memory(512);
        if (http->protocol != NULL)
            stm_printf(http->headers, "%s %ld %s", tc(http->protocol), (long)code, [rmessage UTF8String]);
        else
            stm_printf(http->headers, "HTTP/1.1 %ld %s", (long)code, [rmessage UTF8String]);

        stm_write_char(http->headers, 13);
        stm_write_char(http->headers, 10);

        for (NSString *key in rheaders)
        {
            NSString *value = [rheaders objectForKey:key];
            stm_printf(http->headers, "%s: %s", [key UTF8String], [value UTF8String]);
            stm_write_char(http->headers, 13);
            stm_write_char(http->headers, 10);
        }

        http->body = stm_memory(dsize + 32);
        stm_write(http->body, dbytes, dsize);
        http->error = ekIOK;
        http->response = TRUE;
    }
    else
    {
        http->error = ekIUNDEF;
        http->response = TRUE;
    }

#if defined(MAC_OS_X_VERSION_10_9) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_9
    bmutex_unlock(http->mutex);
#endif
}

/*---------------------------------------------------------------------------*/

static void i_request(OSHttp *http, NSString *verb, const char_t *path, const byte_t *data, const uint32_t size, const bool_t auto_redirect, ierror_t *error)
{
    cassert_no_null(http);
    http->response = FALSE;
    str_destopt(&http->protocol);
    ptr_destopt(stm_close, &http->headers, Stream);
    ptr_destopt(stm_close, &http->body, Stream);
    http->error = ekIOK;

    {
        NSString *curl = [[NSString alloc] initWithFormat:@"%s/%s", tc(http->host_url), path];
        NSURL *url = [[NSURL alloc] initWithString:curl];
        [http->request setURL:url];
        [url release];
        [curl release];
    }

    [http->request setHTTPMethod:verb];

    if (data != NULL)
    {
        NSData *nsdata = [NSData dataWithBytes:(const void *)data length:(NSUInteger)size];
        [http->request setHTTPBody:nsdata];
    }
    else
    {
        cassert(size == 0);
    }

#if defined(MAC_OS_X_VERSION_10_11) && MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_11
    {
        /* Synchronous request */
        bool_t end = FALSE;
        NSURLSessionTask *task = nil;
        http->delegate->auto_redirect = auto_redirect;
        task = [http->session dataTaskWithRequest:http->request
                                completionHandler:^(NSData *_Nullable ddata, NSURLResponse *_Nullable response, NSError *_Nullable lerror) {
                                  i_response(ddata, response, lerror, http);
                                }];

        [task resume];

        while (!end)
        {
            bmutex_lock(http->mutex);
            end = http->response;
            bmutex_unlock(http->mutex);
        }
    }
#else
    {
        NSURLResponse *response = nil;
        NSError *lerror = nil;
        NSData *ddata = [NSURLConnection sendSynchronousRequest:http->request returningResponse:&response error:&lerror];
        unref(auto_redirect);
        i_response(ddata, response, lerror, http);
    }
#endif

    ptr_assign(error, http->error);
}

/*---------------------------------------------------------------------------*/

void oshttp_get(OSHttp *http, const char_t *path, const byte_t *data, const uint32_t size, const bool_t auto_redirect, ierror_t *error)
{
    i_request(http, @"GET", path, data, size, auto_redirect, error);
}

/*---------------------------------------------------------------------------*/

void oshttp_post(OSHttp *http, const char_t *path, const byte_t *data, const uint32_t size, const bool_t auto_redirect, ierror_t *error)
{
    i_request(http, @"POST", path, data, size, auto_redirect, error);
}

/*---------------------------------------------------------------------------*/

Stream *oshttp_response(OSHttp *http)
{
    Stream *stm = NULL;
    cassert_no_null(http);
    cassert(http->response == TRUE);
    if (http->error == ekIOK)
    {
        stm = http->headers;
        http->headers = NULL;
    }

    return stm;
}

/*---------------------------------------------------------------------------*/

void oshttp_response_body(OSHttp *http, Stream *body, ierror_t *error)
{
    cassert_no_null(http);
    cassert(http->response == TRUE);

    if (http->error == ekIOK)
    {
        uint32_t size = stm_buffer_size(http->body);
        stm_pipe(http->body, body, size);
    }

    ptr_assign(error, http->error);
}

File: extensions/libdepwin/nappgui/src/inet/linux/oshttpreq.c
/*
 * NAppGUI Cross-platform C SDK
 * 2015-2024 Francisco Garcia Collado
 * MIT Licence
 * https://nappgui.com/en/legal/license.html
 *
 * File: oshttpreq.c
 *
 */

/* HTTP request (LibCURL-based implementation) */

#include "oshttpreq.inl"
#include <core/arrst.h>
#include <core/heap.h>
#include <core/stream.h>
#include <core/strings.h>
#include <sewer/cassert.h>
#include <sewer/ptr.h>
#include <curl/curl.h>

#if !defined(__LINUX__)
#error This file is only for Linux system
#endif

struct _oshttp_t
{
    CURL *curl;
    struct curl_slist *headers;
    String *host_url;
    bool_t secure;
    Stream *resp_headers;
    Stream *resp_data;
    ierror_t error;
};

/*---------------------------------------------------------------------------*/

void oshttp_init(void)
{
    curl_global_init(CURL_GLOBAL_ALL);
}

/*---------------------------------------------------------------------------*/

void oshttp_finish(void)
{
    curl_global_cleanup();
}

/*---------------------------------------------------------------------------*/

OSHttp *oshttp_create(const char_t *host, const uint16_t port, const bool_t secure)
{
    OSHttp *http = heap_new(OSHttp);
    http->curl = curl_easy_init();
    http->error = ekIOK;
    http->secure = secure;
    http->headers = NULL;
    http->resp_headers = NULL;
    http->resp_data = NULL;

    if (secure == TRUE)
        http->host_url = str_printf("https://%s", host);
    else
        http->host_url = str_printf("http://%s", host);

    if (http->curl != NULL)
    {
        int res = curl_easy_setopt(http->curl, CURLOPT_PORT, port);
        cassert_unref(res == CURLE_OK, res);
    }
    else
    {
        http->error = ekIUNDEF;
    }

    return http;
}

/*---------------------------------------------------------------------------*/

void oshttp_destroy(OSHttp **http)
{
    cassert_no_null(http);
    cassert_no_null(*http);

    if ((*http)->headers != NULL)
    {
        curl_slist_free_all((*http)->headers);
        (*http)->headers = NULL;
    }

    str_destroy(&(*http)->host_url);

    if ((*http)->curl != NULL)
        curl_easy_cleanup((*http)->curl);

    if ((*http)->resp_headers != NULL)
        stm_close(&(*http)->resp_headers);

    if ((*http)->resp_data != NULL)
        stm_close(&(*http)->resp_data);

    heap_delete(http, OSHttp);
}

/*---------------------------------------------------------------------------*/

void oshttp_clear_headers(OSHttp *http)
{
    cassert_no_null(http);
    if (http->headers != NULL)
    {
        curl_slist_free_all(http->headers);
        http->headers = NULL;
    }
}

/*---------------------------------------------------------------------------*/

void oshttp_add_header(OSHttp *http, const char_t *name, const char_t *value)
{
    String *str = NULL;
    cassert(http != NULL);
    str = str_printf("%s: %s", name, value);
    http->headers = curl_slist_append(http->headers, tc(str));
    str_destroy(&str);
}

/*---------------------------------------------------------------------------*/

static size_t i_write_response(char *buffer, size_t size, size_t nitems, void *userdata)
{
    Stream *stm = (Stream *)userdata;
    stm_write(stm, (const byte_t *)buffer, (uint32_t)(size * nitems));
    return nitems * size;
}

/*---------------------------------------------------------------------------*/

static void i_request(OSHttp *http, const bool_t use_get, const char_t *path, const byte_t *data, const uint32_t size, const bool_t auto_redirect, ierror_t *error)
{
    int res = 0;
    cassert_no_null(http);

    if (http->error != ekIOK)
    {
        ptr_assign(error, http->error);
        return;
    }

    /* Seems that CURLOPT_FOLLOWLOCATION fails */
    res = curl_easy_setopt(http->curl, CURLOPT_FOLLOWLOCATION, auto_redirect ? 1L : 0L);
    cassert_unref(res == CURLE_OK, res);

    {
        String *url = str_printf("%s%s", tc(http->host_url), path);
        res = curl_easy_setopt(http->curl, CURLOPT_URL, tc(url));
        cassert_unref(res == CURLE_OK, res);
        str_destroy(&url);
    }

    if (http->headers != NULL)
    {
        res = curl_easy_setopt(http->curl, CURLOPT_HTTPHEADER, http->headers);
        cassert_unref(res == CURLE_OK, res);
    }

    if (data != NULL)
    {
        res = curl_easy_setopt(http->curl, CURLOPT_POSTFIELDSIZE, size);
        cassert_unref(res == CURLE_OK, res);
        res = curl_easy_setopt(http->curl, CURLOPT_POSTFIELDS, (char *)data);
        cassert_unref(res == CURLE_OK, res);
    }

    if (use_get == TRUE)
    {
        res = curl_easy_setopt(http->curl, CURLOPT_HTTPGET, 1);
        cassert_unref(res == CURLE_OK, res);
    }
    else
    {
        res = curl_easy_setopt(http->curl, CURLOPT_POST, 1);
        cassert_unref(res == CURLE_OK, res);
    }

    if (http->resp_headers != NULL)
        stm_close(&http->resp_headers);

    if (http->resp_data != NULL)
        stm_close(&http->resp_data);

    http->resp_headers = stm_memory(100 * 1024);
    http->resp_data = stm_memory(1024 * 1024);

    res = curl_easy_setopt(http->curl, CURLOPT_HEADERFUNCTION, i_write_response);
    cassert(res == CURLE_OK);
    res = curl_easy_setopt(http->curl, CURLOPT_HEADERDATA, http->resp_headers);
    cassert(res == CURLE_OK);

    res = curl_easy_setopt(http->curl, CURLOPT_WRITEFUNCTION, i_write_response);
    cassert(res == CURLE_OK);
    res = curl_easy_setopt(http->curl, CURLOPT_WRITEDATA, http->resp_data);
    cassert(res == CURLE_OK);

    res = curl_easy_perform(http->curl);
    if (res == CURLE_OK)
    {
        ptr_assign(error, ekIOK);
    }
    else
    {
        /* TODO: Error codes */
        ptr_assign(error, ekISERVER);
    }
}

/*---------------------------------------------------------------------------*/

void oshttp_get(OSHttp *http, const char_t *path, const byte_t *data, const uint32_t size, const bool_t auto_redirect, ierror_t *error)
{
    i_request(http, TRUE, path, data, size, auto_redirect, error);
}

/*---------------------------------------------------------------------------*/

void oshttp_post(OSHttp *http, const char_t *path, const byte_t *data, const uint32_t size, const bool_t auto_redirect, ierror_t *error)
{
    i_request(http, FALSE, path, data, size, auto_redirect, error);
}

/*---------------------------------------------------------------------------*/

Stream *oshttp_response(OSHttp *http)
{
    Stream *stm = NULL;
    cassert_no_null(http);
    stm = http->resp_headers;
    http->resp_headers = NULL;
    return stm;
}

/*---------------------------------------------------------------------------*/

void oshttp_response_body(OSHttp *http, Stream *body, ierror_t *error)
{
    const byte_t *data = NULL;
    uint32_t size = 0;
    cassert_no_null(http);
    data = stm_buffer(http->resp_data);
    size = stm_buffer_size(http->resp_data);
    stm_write(body, data, size);
    stm_close(&http->resp_data);
    ptr_assign(error, ekIOK);
}
